pytorch_lightning/callbacks/model_checkpoint.py

```diff
@@ -144,7 +149,7 @@ def __init__(
         save_last: Optional[bool] = None,
         save_top_k: Optional[int] = None,
         save_weights_only: bool = False,
-        mode: str = "auto",
+        mode: str = "min",
```

Comment:
If `mode` is set to `"min"` by default in this release, former codebase that didn't register the `mode` (use the default kwargs, `"default"`) won't be notified about the deprecation, and that might cause problems (e.g in case the "auto" mode inferred to use "max").

Reply:
we can keep the auto functionality and during the deprecation phase print the warning that mode will become "min" by default in future.

Reply:
I think that would be a better deprecation process.

Reply:
The drawback is that this warning would show up for every user, even if they don't configure any custom callback. Since the Trainer configures a default checkpoint callback (with mode auto) it could be confusing to users not knowing about the checkpoint callback. 
A way around this could be to have an internal flag `_show_deprecation_warning` on the checkpoint callback and in the Trainer we set it to false when configuring the default one.

Reply:
Or define this particular model checkpoint in the min mode.

Reply:
how about when it's set to True, we can set the mode to something else and treat it as ``auto`` mode until it gets removed? maybe `default-auto` or `dep-auto` or something else?? and not raise the warning in this case.

Reply:
> The drawback is that this warning would show up for every user, even if they don't configure any custom callback. Since the Trainer configures a default checkpoint callback (with mode auto) it could be confusing to users not knowing about the checkpoint callback.

The default checkpoint callback uses `mode='auto'` but also `monitor=None`. That means it actually defaults to `min` due to this condition:

https://github.com/PyTorchLightning/pytorch-lightning/blob/f677efe61ecf28af1e5216a5baf502be6f394564/pytorch_lightning/callbacks/model_checkpoint.py#L314-L316

So, to keep backwards compatibility and to show a warning only when appropriate, we should keep `mode='auto'` in the ModelCheckpoint signature but `mode='min'` here:

https://github.com/PyTorchLightning/pytorch-lightning/blob/f677efe61ecf28af1e5216a5baf502be6f394564/pytorch_lightning/trainer/connectors/callback_connector.py#L74

This way, users who used a custom ModelCheckpoint without setting the mode would get the warning but users who used the default one wouldn't.

Reply:
Aha, good observation. That makes it much easier:)